File: LinkedList/leetcode148.java
import java.util.List;
import java.util.LinkedList;
import java.util.ArrayList;
import java.util.Arrays;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
if(head == null){
    return null;
}

int cnt =0;
ListNode node = head;




while(node !=null){
    cnt++;
    node = node.next;
}


int[] arr = new int [cnt];

node = head;

for(int i=0; i<cnt; i++){
    arr[i] = node.val;
    node = node.next;
}


Arrays.sort(arr);

ListNode dummyHead = new ListNode(0);
ListNode current = dummyHead;




for(int val: arr){
    current.next = new ListNode(val);
    current = current.next;
}


return dummyHead.next;

        
    }
}

File: LinkedList/leetcode148.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # exception branch
        if not head or not head.next:
            return head
    
        slow, fast = head, head.next
        while fast and fast.next :
            slow = slow.next
            fast = fast.next.next

        mid = slow 
        right = slow.next
        mid.next = None
        left_sorted = self.sortList(head)
        right_sorted = self.sortList(right)

        return self.merge(left_sorted , right_sorted)
    
    # merge sort -> recursion 
    def merge(self, l1 , l2):
        dummy = ListNode(0)
        curr = dummy
        
        while l1 and l2:
            if l1.val <= l2.val:
                curr.next = l1
                l1 = l1.next

            else:
                curr.next = l2
                l2 = l2.next
            
            curr = curr.next
        
        if l1 :
            curr.next = l1
        else :
            curr.next = l2

        return dummy.next
